### ISSUE TITLE
[CN] Align `unchanged` syntax with other CN syntax 

### ISSUE THREAD

**Issue by septract** [rems-project/cerberus#468]
The `{expr} unchanged` keyword is an anomaly in the CN syntax. As far as I can tell, no other construct uses either: 

* a prefix argument position, or 
* argument surrounded by curly braces

It seems not very user-friendly to have a distinct syntax in CN for a single special case, rarely used in practice. I think it'd be better to align `unchanged` with the look and structure of other CN constructs. 

(Looking at the [POPL'23 CN paper](https://dl.acm.org/doi/pdf/10.1145/3571194), it seems like the current design is the remnant of two previous annotations `{expr}@start` and `{expr}@end` that have been removed from the language. I think that design made more sense - if we had multiple constructs that talked about the temporal properties of expressions, it would be okay to have a special syntax.) 

Here's an example of the current syntax: 
```
void unchanged_curr(int n)
/*@ requires 0i32 < n; @*/
{
  int i = 0;
  while (i < n)
  /*@ inv 0i32 <= i; 
          i <= n;
          {n}unchanged; @*/ // <-- current syntax 
  {
    i = i + 1;
  };
}
```

## Proposed resolution: make `unchanged` look like a regular predicate 

One way to make `unchanged` more regular would be to convert it into a regular predicate. So we would write: 
```
void unchanged_new_1(int n)
/*@ requires 0i32 < n; @*/
{
  int i = 0;
  while (i < n)
  /*@ inv 0i32 <= i; 
          i <= n;
          unchanged(n); @*/ // <-- new syntax 
  {
    i = i + 1;
  };
}
``` 

The downside of doing this is that `unchanged` is absolutely not a standard predicate, in that it changes the interpretation of the expression it takes as an argument. But I think that confusion is already potentially a problem, and the current syntax for `unchanged` doesn't do much to help set the correct expectations. 

I think if we only go part way, it's more important to change the syntax so `unchanged` takes a suffix argument, than it is to change the curly brackets. 

## Alternative suggestions: 

* We could define *more* temporal keywords. Eg we could introduce a per-loop version of `unchanged`, or reintroduce `{expr}@start` and `{expr}@end`. This might let us define `unchanged` as a standard predicate, and might also let us resolve the semantic confusion discussed in #443. 
* We could define `unchanged` as a specification-level item, the equivalent of `requires` / `ensures`. Eg. we would write: 

```
void unchanged_new_2(int n)
/*@ requires 0i32 < n; @*/
{
  int i = 0;
  while (i < n)
  /*@ unchanged n; // <-- alt new syntax 
      inv 0i32 <= i; 
          i <= n; @*/ 
  {
    i = i + 1;
  };
}
```

**Comment by cp526** [comment]
I agree that the syntax you propose is better. But there's a misunderstandings of the current state of CN, I think. We do still have the `{...}@start` and `{...}@end` syntax. And that means there is no special magic about `{e} unchanged` (it doesn't change the interpretation of `e`): it's a standard predicate that just desugars into `e == {e}@start`. 

The `unchanged` keyword can be used not just on variables, but also on whole expressions (`{e}unchanged` for an arbitrary e, similarly `{e}@start` and `{e}@end`). It can also be embedded as a subexpression within other expressions; promoting `unchanged` to be a spec-level keyword would not support that.

Either way, perhaps instead of fixing the syntax we should just retire `unchanged` altogether (see #472).
